include/pybind11/pybind11.h

```diff
@@ -1075,15 +1083,27 @@ class class_ : public detail::generic_type {
     }
 
     /// Uses return_value_policy::reference_internal by default
+    template <typename Getter, typename Setter, typename... Extra>
+    class_ &def_property(const char *name, const Getter &fget, const Setter &fset, const Extra& ...extra) {
+        return def_property(name,
+                cpp_function(fget, is_method<type>(*this)), cpp_function(fset, is_method<type>(*this)),
+                return_value_policy::reference_internal, extra...);
+    }
```

Comment:
You've ended up with 4 versions of def_property, for the four combination of {cpp_function, generic} for fget and for fset. Possibly the code would be simpler given a helper template function that takes a getter/setter and a (list of) default call policy, and wraps it into a cpp_function with `is_method` if it isn't already a cpp_function, or returns the provided cpp_function as-is if it is. I think that would allow for only a single version of `def_property`, and probably also simplifies `def_property_readonly`.

By the way, is there any particular reason that def and def_static take their function via perfect forwarding while def_property takes it by const reference?